### PR TITLE
crypto modules: Improve requirements / error messages

### LINKED ISSUE #53140: Incorrect error message from openssl_publickey
##### SUMMARY

I was attempting to make a playbook which would ensures that a target server (all RHEL 7 in this situation) has a key & CSR. openssl_privatekey runs but openssl_publickey fails with the message “the python pyOpenSSL module is required” despite it being installed

##### ISSUE TYPE
- Bug Report

##### ANSIBLE VERSION
```
ansible 2.7.8
  config file = /Users/cadams/Projects/lc-prometheus-server/ansible.cfg
  configured module search path = ['/Users/cadams/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/Cellar/ansible/2.7.8/libexec/lib/python3.7/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 3.7.2 (default, Feb 12 2019, 08:15:36) [Clang 10.0.0 (clang-1000.11.45.5)]
```

##### CONFIGURATION

```
DEFAULT_SCP_IF_SSH(/Users/cadams/Projects/lc-prometheus-server/ansible.cfg) = False
DEFAULT_VAULT_PASSWORD_FILE(/Users/cadams/Projects/lc-prometheus-server/ansible.cfg) = /Users/cadams/Projects/lc-prometheus-server/get_vault_password_from_keyring
RETRY_FILES_ENABLED(/Users/cadams/Projects/lc-prometheus-server/ansible.cfg) = False
```

##### OS / ENVIRONMENT
RHEL 7

##### STEPS TO REPRODUCE
```yaml
- name: Ensure that we have a TLS key for nginx
  block:
      - yum: name=pyOpenSSL state=installed
      - openssl_privatekey:
            path: "/etc/pki/tls/private/{{ inventory_hostname }}.key"
            type: RSA
            size: 4096
      - openssl_publickey:
            path: "/etc/pki/tls/certs/{{ inventory_hostname }}.crt"
            privatekey_path: "/etc/pki/tls/private/{{ inventory_hostname }}.key"
      - openssl_csr:
            path: "/etc/pki/tls/certs/{{ inventory_hostname }}.csr"
            privatekey_path: "/etc/pki/tls/private/{{ inventory_hostname }}.key"
            ocsp_must_staple: true
            common_name: "prometheus.{{ ansible_domain }}"
            subject_alt_name:
                - "{{ inventory_hostname }}"
```

##### EXPECTED RESULTS

A public key file would exist in the expected location

##### ACTUAL RESULTS

```The full traceback is:
WARNING: The below traceback may *not* be related to the actual failure.
  File "/tmp/ansible_openssl_publickey_payload_rw3SSJ/__main__.py", line 131, in <module>
    from cryptography.hazmat.backends import default_backend

fatal: [HOSTNAME]: FAILED! => {
    "changed": false,
    "invocation": {
        "module_args": {
            "attributes": null,
            "backup": null,
            "content": null,
            "delimiter": null,
            "directory_mode": null,
            "follow": false,
            "force": false,
            "format": "PEM",
            "group": null,
            "mode": null,
            "owner": null,
            "path": "/etc/pki/tls/certs/HOSTNAME.crt",
            "privatekey_passphrase": null,
            "privatekey_path": "/etc/pki/tls/private/HOSTNAME.key",
            "regexp": null,
            "remote_src": null,
            "selevel": null,
            "serole": null,
            "setype": null,
            "seuser": null,
            "src": null,
            "state": "present",
            "unsafe_writes": null
        }
    },
    "msg": "the python pyOpenSSL module is required"
}
```

Since Ansible doesn't provide real diagnostic output it's not obvious what the real error was. I had no trouble importing OpenSSL or pip-installing `cryptography` and using that.

### PR DESCRIPTION
**Pull request by felixfontein**
##### SUMMARY
Improves error messages when library cannot be found. Fixes #53140.

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
get_certificate
openssl_certificate
openssl_certificate_info
openssl_csr
openssl_csr_info
openssl_privatekey
openssl_privatekey_info
openssl_publickey

### CODE REVIEW

File: lib/ansible/modules/crypto/openssl_publickey.py [hunk 1]
```diff
@@ -22,7 +22,8 @@
     - Keys are generated in PEM format.
     - This module works only if the version of PyOpenSSL is recent enough (> 16.0.0).
 requirements:
-    - python-pyOpenSSL
+    - pyOpenSSL
+    - cryptography (if I(format) is C(OpenSSL))
```
Comment (MarkusTeufelberger) [inline lib/ansible/modules/crypto/openssl_publickey.py]:
```suggestion
    - cryptography (if I(format) is C(OpenSSH))
```

...right?

File: lib/ansible/modules/crypto/get_certificate.py [hunk 2]
```diff
@@ -44,8 +44,7 @@
   - When using ca_cert on OS X it has been reported that in some conditions the validate will always succeed.
 
 requirements:
-  - "python >= 2.6"
-  - "python-pyOpenSSL >= 0.15"
+  - "pyOpenSSL >= 0.15"
```
Comment (MarkusTeufelberger) [inline lib/ansible/modules/crypto/get_certificate.py:47]: Personally I'd much prefer if al these dependencies were URLs linking to PyPi or some project homepage, but that would be ideally an Ansible-wide change.

File: lib/ansible/modules/crypto/openssl_publickey.py (hunk 1 above)
Comment (felixfontein) [reply lib/ansible/modules/crypto/openssl_publickey.py]: Indeed! Thanks!

File: lib/ansible/modules/crypto/get_certificate.py (hunk 2 above)
Comment (felixfontein) [reply lib/ansible/modules/crypto/get_certificate.py:47]: Yes, that would indeed be nice. Do you want to create a feature request issue for that? (If you do, please CC me.)

**Comment by felixfontein** [comment]
There's currently a problem with the openSUSE servers, which causes the test failures.

**Comment by felixfontein** [comment]
Tests pass here, too.

**Comment by felixfontein** [comment]
Has anyone some opinions on this one, or wants to review it? It would be nice to get it done before the merge window for 2.8.2 closes (i.e. next Monday).

**Comment by MarkusTeufelberger** [comment]
Looks good.

shipit

**Review by resmo — APPROVED** [review]
shipit

**Comment by felixfontein** [comment]
@MarkusTeufelberger @resmo thanks for reviewing and merging!

**Pull request merged**